File: lib/js/test/test-skip-depth.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

QUnit.module('Protocol skip depth limit');

QUnit.test('skip throws DEPTH_LIMIT when depth exceeds 64', function(assert) {
    var transport = new Thrift.Transport('/service');
    var protocol = new Thrift.Protocol(transport);

    assert.throws(
        function() { protocol.skip(Thrift.Type.STRUCT, 64); },
        function(e) {
            return e instanceof Thrift.TProtocolException &&
                   e.type === Thrift.TProtocolExceptionType.DEPTH_LIMIT;
        },
        'skip throws TProtocolException with DEPTH_LIMIT at depth 65'
    );
});

QUnit.test('skip does not throw below the depth limit', function(assert) {
    var transport = new Thrift.Transport('/service');
    var protocol = new Thrift.Protocol(transport);

    // depth=63 → increments to 64, which is not > 64, so no throw before read
    // I32 is a leaf: one readI32 call, no recursion
    transport.setRecvBuffer('\x00\x00\x00\x00');
    assert.ok(protocol.skip(Thrift.Type.I32, 63) !== undefined || true,
        'skip at depth 63 does not throw');
});

// Round-trip the recursion-depth limit through the generated read/write code
// (RecTree struct, RecUnion union and RecError exception from
// JsRecursionDepthTest.thrift) rather than exercising the protocol counter in
// isolation.
QUnit.module('Protocol recursion depth limit');

(function() {
    var LIMIT = Thrift.DEFAULT_RECURSION_DEPTH;
    var DEPTH_LIMIT = Thrift.TProtocolExceptionType.DEPTH_LIMIT;
    var WRITE = Symbol.for('write');
    var READ = Symbol.for('read');

    // Build a linear chain "depth" levels deep. A union needs exactly one field
    // set, so inner nodes set "children" and the innermost one sets its leaf.
    function makeChain(Type, depth, leafField) {
        var leafArgs = {};
        leafArgs[leafField] = 1;
        var node = new Type(leafArgs);
        for (var i = 1; i < depth; i++) {
            node = new Type({children: [node]});
        }
        return node;
    }

    function chainDepth(node) {
        var d = 0;
        while (node) {
            d++;
            node = (node.children && node.children.length) ? node.children[0] : null;
        }
        return d;
    }

    // Serialize through the generated writer. When "unbounded" is set the guard
    // is neutralised on this protocol instance only, so we can craft an
    // over-limit payload and feed it back into a fresh (bounded) reader.
    function serialize(obj, unbounded) {
        var transport = new Thrift.Transport('/service');
        var protocol = new Thrift.Protocol(transport);
        if (unbounded) {
            protocol.incrementRecursionDepth = function() {};
            protocol.decrementRecursionDepth = function() {};
        }
        protocol.writeMessageBegin('', 0, 0);
        obj[WRITE](protocol);
        protocol.writeMessageEnd();
        return transport.send_buf;
    }

    function deserialize(serialized, Type) {
        var transport = new Thrift.Transport('/service');
        transport.setRecvBuffer(serialized);
        var protocol = new Thrift.Protocol(transport);
        protocol.readMessageBegin();
        var obj = new Type();
        obj[READ](protocol);
        protocol.readMessageEnd();
        return obj;
    }

    function isDepthLimit(e) {
        return e instanceof Thrift.TProtocolException && e.type === DEPTH_LIMIT;
    }

    [['RecTree', RecTree, 'item'], ['RecUnion', RecUnion, 'leaf'], ['RecError', RecError, 'leaf']].forEach(function(tc) {
        var tname = tc[0], Type = tc[1], leafField = tc[2];

        QUnit.test(tname + ' round-trips a chain at the depth limit', function(assert) {
            var decoded = deserialize(serialize(makeChain(Type, LIMIT, leafField)), Type);
            assert.equal(chainDepth(decoded), LIMIT,
                'decoded chain has the same depth as the original');
        });

        QUnit.test(tname + ' writing past the depth limit throws DEPTH_LIMIT', function(assert) {
            assert.throws(
                function() { serialize(makeChain(Type, LIMIT + 5, leafField)); },
                isDepthLimit,
                'write throws TProtocolException with DEPTH_LIMIT');
        });

        QUnit.test(tname + ' reading past the depth limit throws DEPTH_LIMIT', function(assert) {
            var serialized = serialize(makeChain(Type, LIMIT + 5, leafField), true);
            assert.throws(
                function() { deserialize(serialized, Type); },
                isDepthLimit,
                'read throws TProtocolException with DEPTH_LIMIT');
        });
    });
})();

File: lib/javame/test/TestSkipDepth.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.protocol.TProtocolException;
import org.apache.thrift.protocol.TProtocolUtil;
import org.apache.thrift.protocol.TType;
import org.apache.thrift.transport.TMemoryBuffer;

/**
 * TProtocolUtil.skip() is the path an unknown field takes, so its nesting is
 * chosen by the peer rather than by the IDL. It threads a maxDepth and
 * decrements it per level, which reads as bounded -- but it was seeded with
 * Integer.MAX_VALUE, so the guard could not fire and the only thing ending the
 * recursion was the stack running out. That arrives as StackOverflowError,
 * an Error rather than a TException, and escapes every handler expecting one.
 *
 * TProtocol already carried DEFAULT_RECURSION_DEPTH for the generated
 * read/write path (THRIFT-6055); skipping descends the same way reading does,
 * so it draws on the same budget.
 *
 * JavaME has no build harness; from the lib/javame directory run:
 *
 *   javac -d /tmp/jme $(find src -name '*.java' ! -name 'THttpClient.java') \
 *       test/TestSkipDepth.java
 *   java -cp /tmp/jme TestSkipDepth
 *
 * (THttpClient.java is excluded because it needs the javax.microedition API.)
 */
public class TestSkipDepth {

  private static int checks = 0;
  private static int failures = 0;

  private static void check(boolean ok, String what) {
    checks++;
    if (ok) {
      System.out.println("ok   - " + what);
    } else {
      failures++;
      System.out.println("FAIL - " + what);
    }
  }

  /**
   * Writes `depth` nested structs, each holding one field of type STRUCT, then
   * closes them all. Every level costs the peer two bytes.
   */
  private static TMemoryBuffer nested(int depth) throws Exception {
    TMemoryBuffer buffer = new TMemoryBuffer(64 + depth * 4);
    TProtocol proto = new TBinaryProtocol(buffer);
    for (int i = 0; i < depth; i++) {
      proto.writeByte(TType.STRUCT);
      proto.writeI16((short) 1);
    }
    // One STOP ends the innermost struct's field list, then one more for each
    // struct on the way back out.
    for (int i = 0; i <= depth; i++) {
      proto.writeByte(TType.STOP);
    }
    return buffer;
  }

  private static void skipNested(int depth) throws Exception {
    TProtocol proto = new TBinaryProtocol(nested(depth));
    TProtocolUtil.skip(proto, TType.STRUCT);
  }

  public static void main(String[] args) throws Exception {
    check(TProtocol.DEFAULT_RECURSION_DEPTH == 64,
        "the shared recursion limit is 64");

    {
      boolean depthLimit = false;
      boolean stackOverflow = false;
      try {
        // Two bytes per level, so this is about 200 KB on the wire.
        skipNested(100000);
      } catch (TProtocolException e) {
        depthLimit = e.getType() == TProtocolException.DEPTH_LIMIT;
      } catch (StackOverflowError e) {
        stackOverflow = true;
      }
      check(depthLimit,
          "deep nesting is refused with DEPTH_LIMIT");
      check(!stackOverflow,
          "deep nesting does not reach StackOverflowError");
    }

    {
      boolean ok = true;
      try {
        skipNested(TProtocol.DEFAULT_RECURSION_DEPTH - 1);
      } catch (Throwable t) {
        ok = false;
        System.out.println("      (" + t + ")");
      }
      check(ok, "nesting within the limit is still skipped");
    }

    {
      boolean depthLimit = false;
      try {
        skipNested(TProtocol.DEFAULT_RECURSION_DEPTH + 2);
      } catch (TProtocolException e) {
        depthLimit = e.getType() == TProtocolException.DEPTH_LIMIT;
      }
      check(depthLimit, "nesting just past the limit is refused");
    }

    System.out.println("\n" + checks + " checks, " + failures + " failures");
    if (failures > 0) {
      System.exit(1);
    }
  }
}
